### PR TITLE
feat: add support to use a different KMS key for backup encryption

### PR DESCRIPTION
**Pull request by jor2**
### Description

[Add support to use a different KMS key for backup encryption](https://github.com/terraform-ibm-modules/terraform-ibm-icd-elasticsearch/issues/323)
[Expose the ability to provision from a backup CRN in the Elasticsearch DA](https://github.com/terraform-ibm-modules/terraform-ibm-icd-elasticsearch/issues/322)

### Release required?

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [x] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

##### Release notes content

[Add support to use a different KMS key for backup encryption](https://github.com/terraform-ibm-modules/terraform-ibm-icd-elasticsearch/issues/323)

[Expose the ability to provision from a backup CRN in the Elasticsearch DA](https://github.com/terraform-ibm-modules/terraform-ibm-icd-elasticsearch/issues/322)

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

### CODE REVIEW

File: ibm_catalog.json [hunk 1]
```diff
@@ -274,6 +278,9 @@
                 {
                   "key": "auto_scaling"
                 },
+                {
+                  "key": "backup_crn"
+                },
```
Comment (ocofaigh) [inline ibm_catalog.json:283]: what about the other vars you added? `existing_backup_kms_instance_crn` and `existing_backup_kms_key_crn`

File: solutions/standard/main.tf [hunk 2]
```diff
@@ -122,23 +122,95 @@ module "kms" {
   ]
 }
 
+#######################################################################################################################
+# KMS backup encryption key for Postgresql
```
Comment (ocofaigh) [inline solutions/standard/main.tf]: Postgresql -> Elasticsearch

File: solutions/standard/main.tf [hunk 3]
```diff
@@ -122,23 +122,95 @@ module "kms" {
...
+  source  = "terraform-ibm-modules/common-utilities/ibm//modules/crn-parser"
+  version = "1.1.0"
+  crn     = var.existing_backup_kms_instance_crn
+}
+
+resource "ibm_iam_authorization_policy" "backup_kms_policy" {
+  count                       = local.existing_backup_kms_instance_guid == local.existing_kms_instance_guid ? 0 : var.existing_backup_kms_key_crn != null ? 0 : var.existing_backup_kms_instance_crn != null ? !var.skip_iam_authorization_policy ? 1 : 0 : 0
+  provider                    = ibm.kms
+  source_service_account      = local.create_cross_account_auth_policy ? data.ibm_iam_account_settings.iam_account_settings[0].account_id : null
+  source_service_name         = "databases-for-postgresql"
```
Comment (ocofaigh) [inline solutions/standard/main.tf]: wrong source name

File: solutions/standard/main.tf [hunk 4]
```diff
@@ -122,23 +122,95 @@ module "kms" {
...
+  source_resource_group_id    = module.resource_group.resource_group_id
+  target_service_name         = local.backup_kms_service_name
+  target_resource_instance_id = local.existing_backup_kms_instance_guid
+  roles                       = ["Reader"]
+  description                 = "Allow all Postgresql instances in the resource group ${module.resource_group.resource_group_id} to read from the ${local.backup_kms_service_name} instance GUID ${local.existing_backup_kms_instance_guid}"
```
Comment (ocofaigh) [inline solutions/standard/main.tf]: Postgresql -> Elasticsearch

File: solutions/standard/main.tf [hunk 5]
```diff
@@ -122,23 +122,95 @@ module "kms" {
...
+resource "time_sleep" "wait_for_backup_kms_authorization_policy" {
+  depends_on      = [ibm_iam_authorization_policy.backup_kms_policy]
+  create_duration = "30s"
+}
+
+module "backup_kms" {
+  providers = {
+    ibm = ibm.kms
+  }
+  count                       = var.existing_backup_kms_key_crn != null ? 0 : var.existing_backup_kms_instance_crn != null ? 1 : 0
```
Comment (ocofaigh) [inline solutions/standard/main.tf:169]: This logic is missing the `var.use_ibm_owned_encryption_key` (see https://github.com/terraform-ibm-modules/terraform-ibm-icd-postgresql/blob/0003266caee9b9371737941a30e0504676d4d35d/solutions/standard/main.tf#L136C33-L136C65)

**Review by ocofaigh — DISMISSED** [review]
see comments

**Comment by jor2** [comment]
> see comments

fixed, my bad

**Review by shemau — DISMISSED** [review]
The code looks good to me.

There is no backup test that I could, which is not an issue with this PR, so I raised #337.

**Comment by jor2** [comment]
````hcl
 2024/11/14 17:13:31 Terraform apply |     "Result": {
 2024/11/14 17:13:31 Terraform apply |         "details": "{\"error\":\"Unauthorized\",\"description\":\"We were unable to complete your request: Unauthorized. Try again with valid permissions or contact support if the issue persists.\"}",
 2024/11/14 17:13:31 Terraform apply |         "error_code": "RC-ServiceBrokerErrorResponse",
 2024/11/14 17:13:31 Terraform apply |         "message": "Please contact the Service Provider for this error. [401, Unauthorized] We were unable to complete your request: Unauthorized. Try again with valid permissions or contact support if the issue persists.",
 2024/11/14 17:13:31 Terraform apply |         "status_code": 502,
 2024/11/14 17:13:31 Terraform apply |         "transaction_id": "626c-0e77f5656f1a876a"
 2024/11/14 17:13:31 Terraform apply |     },
 2024/11/14 17:13:31 Terraform apply |     "RawResult": null
 2024/11/14 17:13:31 Terraform apply | }
 2024/11/14 17:13:31 Terraform apply | 
 2024/11/14 17:13:31 Terraform apply | 
 2024/11/14 17:13:31 Terraform apply |   with module.elasticsearch[0].module.elasticsearch.ibm_database.elasticsearch,
 2024/11/14 17:13:31 Terraform apply |   on ../../main.tf line 85, in resource "ibm_database" "elasticsearch":
 2024/11/14 17:13:31 Terraform apply |   85: resource "ibm_database" "elasticsearch" {
````

**Review by shemau — DISMISSED** [review]
LGTM

**Review by shemau — APPROVED** [review]
LGTM

**Pull request merged**